Feature: implement get debug-info

10ko opened the pull request:
This PR introduces the command:
```bash
$ garden get debug-info
```
As spec'd out in #547, it's a command to collect all sort of system, project and providers information in a zip file. This will be useful data users can attach to github issues.

Thanks,
E.

edvald commented on docs/reference/commands.md:
```diff
@@ -324,6 +324,25 @@ Outputs the latest execution result of a provided test.
+Examples:
+
+garden get debug-info                # create a zip file on the root of the project with debug information
+garden get debug-info --format yaml  # outputs the provider info as yaml files (default as json)
```
Curious about the motivation for this. Do you have a use case for YAML here?

10ko replied:
We couldn't really agree between us on what's the preferred format (yaml vs json) for the report. Some find yaml more readable others prefer json.
Unless you have some strong opinions regarding this, it seemed a simple feature to add and we added it.

edvald replied:
Just seemed like extra overhead somehow, but all good, no strong opinions.

edvald commented on garden-service/src/actions.ts:
```diff
@@ -409,6 +410,13 @@ export class ActionHelper implements TypeGuard {
+  async getDebugInfo({ log }: { log: LogEntry }): Promise<DebugInfoMap> {
+    const handlers = this.getActionHandlers("getDebugInfo")
+    const res = await Bluebird.props(mapValues(handlers, h => h({ ...this.commonParams(h, log) })))
+    log.setSuccess("Ready")
```
This line might belong in the surrounding context. Could then `return Bluebird.props...` above for brevity.

edvald commented on garden-service/src/commands/get/get-debug-info.ts:
```diff
@@ -0,0 +1,209 @@
+    },
+  }
+  const paths: string[] = []
+  for await (const item of scanDirectory(root, scanOpts)) {
```
It looks like this routine duplicates code from the Garden class. Might be worth factoring out, unless there's a reason it can't be?

edvald replied:
Side note: This async generator thingy isn't actually very nice (that's on me, I made it :P), I or someone willing should refactor it at some point.

10ko replied:
You are right, it was duplicated logic and I extracted it in its own function.
Regarding the generator thingy, I thought of giving it a go but I decided against since this PR is already huge.

edvald commented on garden-service/src/commands/get/get-debug-info.ts:
```diff
@@ -0,0 +1,209 @@
+import { zipFolder } from "../../util/archive"
+
+export const TEMP_DEBUG_ROOT = "tmp"
+export const SYSTEM_INFO_FILENAME = "systemInfo.json"
```
Maybe kebab-case this for consistency?

edvald commented on garden-service/src/commands/get/get-debug-info.ts:
```diff
@@ -0,0 +1,209 @@
+  await zipFolder(tempPath, join(root, outputFilename), log)
+
+  await remove(tempPath)
+  log.info(`Done! Please find your report under ${root}.`)
```
Super minor, but `... at <full path to output file>` would allow me to quickly open or copy the path.

edvald commented on garden-service/src/commands/get/get-debug-info.ts:
```diff
@@ -0,0 +1,209 @@
+  log.info(`Done! Please find your report under ${root}.`)
+}
+
+export async function collectBasicDebugInfo(root: string, log: LogEntry) {
```
Would be great to add docstrings explaining the functions, e.g. how this one should always work even if the debug logs command fails etc.

edvald commented on garden-service/src/plugins/kubernetes/kubernetes.ts:
```diff
@@ -155,6 +160,28 @@ export async function configureProvider({ projectName, config }: ConfigureProvid
+
+  const namespacesList = [appNamespace, appMetadataNamespace, systemNamespace, systemMetadataNamespace]
+  const namespaces = await Bluebird.map(namespacesList, async (ns) => {
+    const out = await kubectl.stdout({ log, context, args: ["get", "all", "--namespace", ns, "--output", "json"] })
```
Might need to catch the case where the namespace doesn't exist.

10ko replied:
I didn't think about that but it looks like if the namespace doesn't exist it won't throw any exception.
Instead the output will look like the following:

```json
{
    "namespace": "trash",
    "output": {
        "apiVersion": "v1",
        "items": [],
        "kind": "List",
        "metadata": {
            "resourceVersion": "",
            "selfLink": ""
        }
    }
},
```

10ko replied:
Is that something we are happy with?

edvald replied:
Ah ok, wasn't sure if kubectl would balk at this. In that case it's all good.

eysi09 commented on garden-service/src/commands/get/get-debug-info.ts:
```diff
@@ -0,0 +1,209 @@
+
+export async function collectSystemDiagnostic(root: string, log: LogEntry) {
+
+  log.info("Collecting OS basic information.")
```
Minor but shouldn't this be, `"Collecting basic OS information"`, or alternatively, just `"Collecting basic system information"`.

eysi09 replied:
...in my last comment I actually suggest skipping this all together and moving most of the logging the `action` body.

eysi09 commented on garden-service/src/commands/get/get-debug-info.ts:
```diff
@@ -0,0 +1,209 @@
+  description = dedent`
+    Examples:
+
+    garden get debug-info                # create a zip file on the root of the project with debug information
```
It should `at` instead of `on` in doc string: `"# create  a zip file at the root of the project...`. Right?

eysi09 commented on garden-service/src/commands/get/get-debug-info.ts:
```diff
@@ -0,0 +1,209 @@
+    Examples:
+
+    garden get debug-info                # create a zip file on the root of the project with debug information
+    garden get debug-info --format yaml  # outputs the provider info as yaml files (default as json)
```
For consistency, use `# output...`  instead of `# outputs...`.

eysi09 commented on garden-service/test/unit/src/commands/get/get-debug-info.ts:
```diff
@@ -0,0 +1,146 @@
+    return fileName.match(debugZipFileRegex)
+  })
+  for await (const name of deleteFilenames) {
+    console.log(join(root, name))
```
Remove `console.log`?

eysi09 commented on garden-service/test/unit/src/commands/get/get-debug-info.ts:
```diff
@@ -0,0 +1,146 @@
+    )
+  })
+
+  describe("generateBasicDebugInfo", () => {
```
Feels like this could be split up into several `it` statements since it has multiple `expect`s. Can be hard to debug if one of them fails.

eysi09 commented on garden-service/src/commands/get/get-debug-info.ts:
```diff
@@ -0,0 +1,209 @@
+  arguments = debugInfoArguments
+  options = debugInfoOptions
+
+  async action({ garden, log, opts }: CommandParams<Args, Opts>) {
```
The command appears to run twice. See e.g. the logs:

```
✔ Ready
Collecting project configuration files.
Collecting OS basic information.
Preparing archive.
Done! Please find your report under /Users/eysi/code/garden-io/garden/examples/simple-project.
Collecting project configuration files.
Collecting OS basic information.
Preparing archive.
Done! Please find your report under /Users/eysi/code/garden-io/garden/examples/simple-project.
```

eysi09 commented on garden-service/src/commands/get/get-debug-info.ts:
```diff
@@ -0,0 +1,209 @@
+export const DEBUG_ZIP_FILENAME = "debug-info-TIMESTAMP.zip"
+export const PROVIDER_INFO_FILENAME_NO_EXT = "info"
+
+export async function generateBasicDebugInfoReport(root: string, log: LogEntry) {
```
Is this function used anywhere?

10ko replied:
Yes, this is the one that should have been called in the `cli.ts` catch block.
I swapped names during my refactoring and that slipped through. Thanks for pointing out.

eysi09 commented:
Had a few nitpicks on my own but otherwise looks great!

One thing we could improve though is the logging. E.g.: the body of the command action could look like this:

```
    const tempPath = join(garden.projectRoot, GARDEN_DIR_NAME, TEMP_DEBUG_ROOT)
    const entry = log.info({ msg: "Collecting debug info", status: "active" })
    const systemEntry = entry.info({ section: "system", msg: "collecting info", status: "active" })
    await collectBasicDebugInfo(garden.projectRoot, systemEntry)
    systemEntry.setSuccess({ msg: chalk.green(`Done (took ${systemEntry.getDuration(1)} sec)`), append: true })

    const providerEntry = entry.info({ section: "providers", msg: "collecting info", status: "active" })
    await collectProviderDebugInfo(garden, providerEntry, tempPath, opts.format)
    providerEntry.setSuccess({ msg: chalk.green(`Done (took ${providerEntry.getDuration(1)} sec)`), append: true })

    entry.setState("Preparing archive")
    const outputFilename = DEBUG_ZIP_FILENAME.replace("TIMESTAMP", new Date().toISOString())
    await zipFolder(tempPath, join(garden.projectRoot, outputFilename), log)

    await remove(tempPath)

    entry.setSuccess({ msg: "Done", append: true })
    log.info(`\nDone! Please find your report under ${garden.projectRoot}.`)

    return { result: 0 }
```

With this, we can simply remove the log statements from `actions.getDebugInfo` (and return the Bluebird map like @edvald suggested) and from `collectSystemDiagnostic`. The info statements that is, no the errors. 

Passing the system and provider entries like this also ensures that the errors get set on the appropriate entry. 

The end result looks something like this:

```
✔ Preparing archive → Done
   ✔ system                    → collecting project configuration files → Done (took 0.3 sec)
   ✔ providers                 → collecting info → Done (took 0.4 sec)

Done! Please find your report under /Users/eysi/code/garden-io/garden/examples/simple-project.
```

10ko commented:
Dear all, 
thanks a lot for the review, I agree on everything and wrote fixes and add-ons as you requested.

I have few notes and consideration to add:

- @eysi09 The reason the command was running twice was because the garden project you tried the command against was:
  * a valid garden project, so
  * the normal command action would get called
  * at which point the command tried to collect the providers info but, more than likely, you had docker or k8s down.
  * the function `collectProviderDebugInfo(...)` would throw an exception. 
  * The exception would then get caught in the `cli.ts` where `generateBasicDebugInfoReport(...)` would get called again: I added another try/catch in the function collecting the provider info.

- The try/catch mentioned above will catch any exception thrown by any provider. Since the actions are executed through `Bluebird.props(...)`: even if one of may providers fail, all (promises) will. It would be nice to catch only the failing providers. This would require a lot of internals work and was kinda out-of-scope for this PR. If any of you think it's something we want in the future I can open an issue.
- @edvald I found a strange 404 error regarding a dependency called `tmp-promise`: the 2.0.0 version we used just got deleted from npm and github. I updated the dependecy to 2.0.1 but maybe we should investigate a replacement or if we need it at all since it doens't seem very reliable?

Thanks again and please let me know if you have any other comment.
E.

eysi09 commented:
I believe the `tmp-promise` issue was resolved in another PR so rebasing should remove the need for that fix.

eysi09 commented:
One thing I noticed after approving is that the debug info is not git ignored.
